File: src/main/java/ir/reservs/reservs/ui/main/history/HistoryFragment.java
package ir.reservs.reservs.ui.main.history;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import javax.inject.Inject;

import ir.reservs.reservs.R;
import ir.reservs.reservs.model.ReserveHistory;
import ir.reservs.reservs.ui.base.BaseFragment;

public class HistoryFragment extends BaseFragment implements HistoryContract.View {
    @Inject
    HistoryAdapter historyAdapter;

    @Inject
    HistoryPresenter historyPresenter;


    private RecyclerView historyRecycler;

    private ProgressBar progressBar;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.layout_history, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        historyPresenter.onAttach(this);
        historyRecycler.setAdapter(historyAdapter);
    }



    @Override
    public void setHistoryData(List<ReserveHistory> reserves) {
        historyAdapter.addItems(reserves);
    }

    @Override
    public void showProgress() {
        progressBar.setVisibility(View.VISIBLE);
        historyRecycler.setVisibility(View.GONE);
    }

    @Override
    public void hideProgress() {
        progressBar.setVisibility(View.GONE);
        historyRecycler.setVisibility(View.VISIBLE);
    }

    @Override
    public void onError(String error) {
        //Snackbar.make(historyRecycler, error, Snackbar.LENGTH_LONG).show();
    }

    @Override
    public void setup(View view) {
        getFragmentComponent().inject(this);
        progressBar = view.findViewById(R.id.progressBar);
        historyRecycler = view.findViewById(R.id.historyRecycler);
    }

    @Override
    public void onError(int resId) {
        onError(getString(resId));
    }

    @Override
    public void onDestroy() {
        if (historyPresenter != null) {
            historyPresenter.onDetach();
            historyPresenter = null;
            progressBar = null;
        }
        super.onDestroy();
    }
}

File: src/main/java/ir/reservs/reservs/ui/main/history/HistoryFragment.kt
package ir.reservs.reservs.ui.main.history

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import ir.reservs.reservs.R
import ir.reservs.reservs.model.History
import ir.reservs.reservs.ui.base.BaseFragment
import ir.reservs.reservs.ui.custome.StateAdapter
import ir.reservs.reservs.ui.main.MainActivity
import ir.reservs.reservs.utils.PaginationScrollListener
import kotlinx.android.synthetic.main.layout_history.*
import javax.inject.Inject

class HistoryFragment : BaseFragment(), HistoryContract.View {
    private var isScrolling = false
    var historyAdapter: HistoryAdapter? = null
        @Inject set

    var historyPresenter: HistoryPresenter? = null
        @Inject set


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.layout_history, container, false)
    }

    override fun setup(view: View) {
        fragmentComponent?.inject(this)
        historyPresenter?.onAttach(this)
        (activity as MainActivity).showBottomNavigation()
        initializeStateAdapter(historyRecycler, historyAdapter as RecyclerView.Adapter<*>)
        getStateAdapter().currentView = StateAdapter.VIEW_EMPTY
        historyRecycler.adapter = getStateAdapter()
        loadingState()
        val llMan = LinearLayoutManager(context)
        historyRecycler.layoutManager = llMan
        historyRecycler.addOnScrollListener(object : PaginationScrollListener(llMan) {
            override fun loadMoreItems() {
                Log.e("loadMoreItems", "fetchHistories called here");
                historyPresenter?.fetchHistories()
            }

            override fun isLastPage(): Boolean {
                Log.e("isLastPage", "isLastPage called")
                return historyPresenter?.isLastPage() ?: false
            }

            override fun isLoading(): Boolean {
                Log.e("isLoading", "isLoading is called")
                return historyPresenter?.isLoading() ?: false
            }

        })
    }

    override fun updateAdapter(it: MutableList<History>) {
        historyAdapter?.addHistory(it)
    }

    override fun onDestroyView() {
        historyPresenter?.onDetach()
        historyPresenter = null
        historyAdapter = null
        super.onDestroyView()
    }

}
